src/conversions/chrono.rs

```diff
@@ -0,0 +1,848 @@
+//!
+//! Unavailable with the `abi3` feature.
+//!
+//! # Setup
+//!
+//! To use this feature, add this to your **`Cargo.toml`**:
+//!
+//! ```toml
+//! [dependencies]
+//! # change * to the latest versions
+//! pyo3 = { version = "*", features = ["chrono"] }
+//! chrono = "0.4"
+// workaround for `extended_key_value_attributes`: https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+#![cfg_attr(docsrs, cfg_attr(docsrs, doc = concat!("pyo3 = { version = \"", env!("CARGO_PKG_VERSION"),  "\", features = [\"chrono\"] }")))]
+#![cfg_attr(
+    not(docsrs),
+    doc = "pyo3 = { version = \"*\", features = [\"chrono\"] }"
+)]
+//! ```
+//!
+//! Note that you must use compatible versions of chrono and PyO3.
+//! The required chrono version may vary based on the version of PyO3.
+//!
+//! # Example: Convert a PyDateTime to chrono's DateTime<Utc>
+//!
+//! ```rust
+//! use chrono::{Utc, DateTime};
+//! use pyo3::{Python, ToPyObject, types::PyDateTime};
+//!
+//! fn main() {
+//!     pyo3::prepare_freethreaded_python();
+//!     Python::with_gil(|py| {
+//!         // Create an UTC datetime in python
+//!         let py_tz = Utc.to_object(py);
+//!         let py_tz = py_tz.cast_as(py).unwrap();
+//!         let pydatetime = PyDateTime::new(py, 2022, 1, 1, 12, 0, 0, 0, Some(py_tz)).unwrap();
+//!         println!("PyDateTime: {}", pydatetime);
+//!         // Now convert it to chrono's DateTime<Utc>
+//!         let chrono_datetime: DateTime<Utc> = pydatetime.extract().unwrap();
+//!         println!("DateTime<Utc>: {}", chrono_datetime);
+//!     });
+//! }
+//! ```
+use crate::exceptions::PyTypeError;
+use crate::types::{
+    timezone_utc, PyDate, PyDateAccess, PyDateTime, PyDelta, PyDeltaAccess, PyTime, PyTimeAccess,
+    PyTzInfo, PyTzInfoAccess,
+};
+use crate::{
+    AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python, ToPyObject,
+};
+use chrono::offset::{FixedOffset, Utc};
+use chrono::{
+    DateTime, Datelike, Duration, NaiveDate, NaiveDateTime, NaiveTime, Offset, TimeZone, Timelike,
+};
+use pyo3_ffi::{PyDateTime_IMPORT, PyTimeZone_FromOffset};
+use std::convert::TryInto;
+
+impl ToPyObject for Duration {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        // Total number of days
+        let days = self.num_days();
+        // Remainder of seconds
+        let secs_dur = *self - Duration::days(days);
+        // .try_into() converts i64 to i32, but this should never overflow
+        // since it's at most the number of seconds per day
+        let secs = secs_dur.num_seconds().try_into().unwrap();
+        // Fractional part of the microseconds
+        let micros = (secs_dur - Duration::seconds(secs_dur.num_seconds()))
+            .num_microseconds()
+            // This should never panic since we are just getting the fractional
+            // part of the total microseconds, which should never overflow.
+            .unwrap()
```

Comment:
I'm not entirely sure whether the plan is still to feedback upstream to `chrono` for possible API additions, this looks like one to me. It could be nice to be able to decompose the `Duration` into some kind of tuple like `(days, seconds, nanos)` where it's known that none of the components overflow, avoiding the need to do the subtraction here ourselves and then `unwrap`.

cc @djc

Reply:
Unfortunately we don't really control the `Duration` API in chrono 0.4, since that's bound by the time 0.1 implementation of `Duration`.

Reply:
IIRC the `TimeDelta` in 0.5 no longer have this issue.

Reply:
Yeah, we can definitely fix this for `TimeDelta` 0.5 (PR to main welcome).